src/test/java/org/jenkinsci/plugins/workflow/steps/durable_task/ShellStepTest.java

```diff
@@ -237,6 +258,97 @@ public DescriptorImpl() {
+        final String credentialsId = "creds";
+        final String username = "bob";
+        final String password = "s3cr3t";
+        UsernamePasswordCredentialsImpl c = new UsernamePasswordCredentialsImpl(CredentialsScope.GLOBAL, credentialsId, "sample", username, password);
+        CredentialsProvider.lookupStores(j.jenkins).iterator().next().addCredentials(Domain.global(), c);
+        j.createSlave("remote", null, null);
+        WorkflowJob p = j.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+            "node('master') {\n" +
+            "  sh 'pwd'\n" +
+            "}\n" +
+            "node('remote') {\n" +
+            "  sh 'pwd'\n" +
+            "  sh 'echo on agent'\n" +
+            "  withCredentials([usernameColonPassword(variable: 'USERPASS', credentialsId: '" + credentialsId + "')]) {\n" +
+            "    sh 'set +x; echo curl -u $USERPASS http://server/'\n" +
+            "  }\n" +
+            "}", true));
+        WorkflowRun b = j.assertBuildStatusSuccess(p.scheduleBuild2(0));
+        j.assertLogContains("+ pwd [master]", b);
+        j.assertLogContains("+ PWD [master → remote]", b);
+        j.assertLogContains("ON AGENT [master → remote]", b);
+        j.assertLogNotContains(password, b);
+        j.assertLogNotContains(password.toUpperCase(Locale.ENGLISH), b);
+        j.assertLogContains("CURL -U **** HTTP://SERVER/ [master → remote]", b);
+    }
+    @TestExtension("remoteLogger") public static class LogFile implements LogStorageFactory {
+        @Override public LogStorage forBuild(FlowExecutionOwner b) {
+            final LogStorage base;
+            try {
+                base = FileLogStorage.forFile(new File(b.getRootDir(), "special.log"));
+            } catch (IOException x) {
+                return new BrokenLogStorage(x);
+            }
+            return new LogStorage() {
+                @Override public BuildListener overallListener() throws IOException, InterruptedException {
+                    return new RemotableBuildListener(base.overallListener());
+                }
+                @Override public TaskListener nodeListener(FlowNode node) throws IOException, InterruptedException {
+                    return new RemotableBuildListener(base.nodeListener(node));
+                }
+                @Override public AnnotatedLargeText<FlowExecutionOwner.Executable> overallLog(FlowExecutionOwner.Executable build, boolean complete) {
+                    return base.overallLog(build, complete);
+                }
+                @Override public AnnotatedLargeText<FlowNode> stepLog(FlowNode node, boolean complete) {
+                    return base.stepLog(node, complete);
+                }
+            };
+        }
+    }
+    private static class RemotableBuildListener implements BuildListener {
+        private static final long serialVersionUID = 1;
+        /** actual implementation */
+        private final TaskListener delegate;
```

Comment:
Would be nice to have another test here too, which would produce a large volume of output, using a listener that did _not_ delegate but wrote independently to a separate file, and use `Channel.Listener.onRead`/`.onWrite` to demonstrate that there is little traffic…unless you disable watch mode, as a control.

Reply:
(resolved)